### PR TITLE
bpo-44682: Handle invalid arg to pdb's "commands" directive

### PR DESCRIPTION
**Pull request by akulakov**
https://bugs.python.org/issue44682

### CODE REVIEW

**Review by auvipy — APPROVED** [review]

File: Misc/NEWS.d/next/Library/2021-07-20-00-11-47.bpo-44682.3m2qVV.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+In :mod:`pdb`, *commands* directive was fixed to disallow setting commands
+for a non-existing breakpoint and to display "out of range" error.
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Library/2021-07-20-00-11-47.bpo-44682.3m2qVV.rst]: It's not always "out of range" - if you look at Bdb.get_bpbynumber there are a number of other possible error messages.

File: Lib/pdb.py [hunk 2]
```diff
@@ -562,6 +562,12 @@ def do_commands(self, arg):
             except:
                 self.error("Usage: commands [bnum]\n        ...\n        end")
                 return
+        try:
+            self.get_bpbynumber(bnum)
+        except ValueError as err:
+            self.error(err)
```
Comment (iritkatriel) [inline Lib/pdb.py]:
```suggestion
            self.error('cannot set commands: %s' % err)
```

File: Lib/test/test_pdb.py [hunk 3]
```diff
@@ -319,6 +320,8 @@ def test_pdb_breakpoint_commands():
     > <doctest test.test_pdb.test_pdb_breakpoint_commands[0]>(5)test_function()
     -> print(3)
     (Pdb) break
+    (Pdb) commands 10
+    *** Breakpoint number 10 out of range
```
Comment (iritkatriel) [inline Lib/test/test_pdb.py]: Try to add tests for the other cases.  https://github.com/python/cpython/blob/cd0c4f358b934322ca024692dde7dfe7cf0b5fb6/Lib/bdb.py#L474

File: Lib/pdb.py (hunk 2 above)
Comment (akulakov) [reply Lib/pdb.py]: As it already displays a more informative error, why do you think a more generic one would work better?

**Comment by akulakov** [comment]
@iritkatriel Thanks for catching this! expanded the tests, fixed news entry & replied to one comment.

File: Lib/pdb.py (hunk 2 above)
Comment (iritkatriel) [reply Lib/pdb.py]: This includes the original error as well, so it's not more generic.  The prefix makes it clear that what you asked for did not happen.

File: Misc/NEWS.d/next/Library/2021-07-20-00-11-47.bpo-44682.3m2qVV.rst [hunk 4]
```diff
@@ -0,0 +1,2 @@
+In :mod:`pdb`, *commands* directive was fixed to disallow setting commands
+for an invalid breakpoint and to display appropriate error.
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Library/2021-07-20-00-11-47.bpo-44682.3m2qVV.rst]:
```suggestion
for an invalid breakpoint and to display an appropriate error.
```

File: Lib/pdb.py (hunk 2 above)
Comment (akulakov) [reply Lib/pdb.py]: you are right, good point :)

**Comment by akulakov** [comment]
@iritkatriel committed your suggestions.

File: Lib/test/test_pdb.py [hunk 5]
```diff
@@ -319,6 +322,14 @@ def test_pdb_breakpoint_commands():
...
+    (Pdb) commands a
+    *** Usage: commands [bnum]
+            ...
+            end
+    (Pdb) commands 4
+    *** Breakpoint 4 already deleted
```
Comment (iritkatriel) [inline Lib/test/test_pdb.py]: I imagine you need to add the prefix to the expected output.

File: Misc/NEWS.d/next/Library/2021-07-20-00-11-47.bpo-44682.3m2qVV.rst [hunk 6]
```diff
@@ -0,0 +1,2 @@
+In :mod:`pdb`, *commands* directive was fixed to disallow setting commands
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Library/2021-07-20-00-11-47.bpo-44682.3m2qVV.rst]:
```suggestion
Change the :mod:`pdb` *commands* directive to disallow setting commands
```

**Comment by iritkatriel** [comment]
It doesn't look related, maybe rebasing will help.

**Comment by iritkatriel** [comment]
@ambv Any idea why the "AMD64 Windows10 PR" test is failing? We tried to rebase, but that didn't seem to help. Thanks.

**Comment by ambv** [comment]
@iritkatriel Currently that particular buildbot worker (bolen-windows10) is failing all builds with:
```
fatal error C1060: compiler is out of heap space
```
I contacted the buildbot owner as this looks like a buildbot failure (Windows 10 Pro PR buildbot is fine, as is Azure Pipelines, and Github Actions).

Unless urgent, I'd recommend waiting for the buildbot to return to a green state. I will let you know on this PR when this happens.

**Comment by ambv** [comment]
Well, the good news is that the Windows 10 buildbot is back and running, the owner notified me it was in a bad state due to Windows updates clogging up most of the RAM.

The bad news is that the Windows 8.1 Refleaks buildbot is red on this PR only. This looks weird to me since the tests passed, just the exit code is non-zero. In any case, I scheduled it to re-run: https://buildbot.python.org/all/#/buildrequests/315077

Unfortunately, refleaks are pretty slow to re-run so we might not be able to see the new result today.

**Comment by akulakov** [comment]
@ambv thanks for following up on this, no rush!

**Comment by akulakov** [comment]
There is one failure that seems to be unrelated AFAICT:

test test_concurrent_futures failed -- Traceback (most recent call last):
  File "D:\buildarea\pull_request.ware-win81-release.refleak\build\lib\test\test_concurrent_futures.py", line 905, in test_idle_thread_reuse
    self.assertEqual(len(executor._threads), 1)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 2 != 1

**Comment by iritkatriel** [comment]
I agree, that doesn't seem related. @ambv any objections to merging this as it is?

**Comment by ambv** [comment]
Thanks for bearing with me on this. I'll look into this importlib weirdness. It's worrying it is reproducible but in any case it is unrelated to this PR.

**Comment by iritkatriel** [comment]
Thank you @ambv!

**Comment by akulakov** [comment]
@ambv No worries, thank you!  @iritkatriel much thanks for reviewing!

**Pull request merged**